### PR TITLE
add tile fetching logic

### PR DESCRIPTION
**Pull request by lovasoa**
This is not my work, but the one of @JohnDoee.

See https://gist.github.com/JohnDoee/6498557f0cd615a6c2845153e5660d68

This is a work in progress

### CODE REVIEW

**Comment by expenses** [comment]
Looks good. Say when you want me to merge.

**Comment by JohnDoee** [comment]
I'll turn it into an actual command-line tool (with some arguments) then I think my part should be good.

@expenses Do you know which algorithm the magic is ? It looks like part of aes-192 (`magic` variable is Inverted S-box, 11 rounds like aes-192, `xor_bytes_by_magic_table` looks like "add round-key"), a few steps are missing though, some shift and mix stuff.

**Comment by lovasoa** [comment]
I didn't know the implementation of AES, but I just had a look at the wikipedia page, and it ideed looks to be the same. I think shift part that you thought was missing is in fact the unusual indexing here : `split_bytes[(y - x) % 4][x]` (https://github.com/lovasoa/gapdecoder/blob/master/decryption.py#L89)

**Comment by lovasoa** [comment]
Here is a python3 AES implementation for comparison: https://gist.github.com/lovasoa/a0b384ca0e9c0a485f212caab68d98ec

**Comment by JohnDoee** [comment]
Cool, I couldn't find a good, readable one. Looks like the xor in the loop is actually inverse mix column.

Also, I tested the round key generation algorithm, it gives correct result if you use the password: 5b63db113b7af3e0b1435556c8f9530c

~~Like the url signing algorithm that's standard + custom stuff, this look like AES-ECB + [custom XORing](https://github.com/expenses/gapdecoder/blob/master/decryption.py#L84-L85)~~

It turns out it is actually AES-CBC with IV 71e70405353a778bfa6fbc30321b9592 - The decrypt only works because you only need to decrypt 32 bytes.

I tested it, it works.

Test string: fe0e0bc5e597932e993b04e1b4be9af60e511d089f65da22cb18c005eca9695f
Result: 0000010501010101010100000000000000000102030405060708090a0bffc400

**Comment by lovasoa** [comment]
~~@JohnDoee I am wondering whether your [`encode_string`](https://github.com/lovasoa/gapdecoder/blob/master/tile_fetch.py#L20-L34) function could not be a simple [PKBF2](https://en.wikipedia.org/wiki/PBKDF2). What do you think ?~~

It is a simple HMAC.

**Comment by JohnDoee** [comment]
This was a lot more simple than I originally thought when looking at the "Arts and Culture" source code. It was a bit eye-opening for me that I didn't really realize the actual algorithms used.

I think these changes should cover most of it.

**Comment by lovasoa** [comment]
@expenses I think we are good to go. Maybe you could add @JohnDoee and me as contributors to this repo ?

File: tile_fetch.py [hunk 1]
```diff
@@ -64,10 +64,9 @@ def load_image_info(url):
 
 def load_tiles(url, z=6):
     tile_info, tile_size, image_name, path, token = load_image_info(url)
-    try:
-        tile = tile_info[z]
-    except IndexError:
-        print('Unknown zoom-level %s' % (z,))
+
+    if z > len(tile_info) or z < 0:
+        print('Invalid zoom level %d. The maximum zoom level is %d' % (z, len(tile_info)))
```
Comment (JohnDoee) [inline tile_fetch.py:69]: The previous solution allowed negative indexes. It'll also throw IndexError later in the code if z == len(tile_info)

My original idea was to allow zoom -1 to always get the best quality.

Comment (lovasoa) [reply tile_fetch.py:69]: Oh, you are right, my mistake. I changed the code to allow negative z and z==len(tile_info). However, I made it so z is always positive after this line. This canonicalizes file names, and allows keeping the real zoom level in the filenames.

**Comment by lovasoa** [comment]
@expenses : Could you open the issues for this project ?

**Pull request merged**